### ISSUE TITLE
Message remains unsettled

### ISSUE THREAD

**Issue by philschleier** [Azure/go-amqp#33]
Hi,

we've been using this library for a while without any issues to connect to a cpp qpid Broker.

The changes introduced with v0.13.6 however break our client.

If I understand correctly, any incoming message is added to `unsettledMessages` via `addUnsettled`, if it has an empty `DeliveryTag`. It should get removed from there by a call to `deleteUnsettled`. 
The only place where `deleteUnsettled` is called is within the `trackCompletion` handler (we're using `HandleMessage`). This however was recently changed to only be run with a receiverSettleMode of ModeSecond.

As can be seen from the following debug log, my link has receiverSettleMode "first", while my message does have a non-empty DeliveryTag "\x00\x00\x00\x00". If I understand correctly, my message gets added to the unsettledMessages map because it has a non-empty DeliveryTag, but because I'm in ModeSecond it never gets removed.

```
19:07:48.210557 FLOW Link Mux half: source: somewhere, inflight: 0, credit: 0, deliveryCount: 0, messages: 0, unsettled: 0, maxCredit : 1, settleMode: first
19:07:48.210628 TX(Session) - tx: Flow{NextIncomingID: 0, IncomingWindow: 100, NextOutgoingID: 0, OutgoingWindow: 100, Handle: 0, DeliveryCount: 0, LinkCredit: 1, Available: <nil>, Drain: false, Echo: false, Properties: map[]}
19:07:51.064771 RX(Session): Transfer{Handle: 0, DeliveryID: 0, DeliveryTag: "\x00\x00\x00\x00", MessageFormat: 0, Settled: false, More: false, ReceiverSettleMode: <nil>, State: <nil>, Resume: false, Aborted: false, Batchable: false, Payload [size]: 1746}
19:07:51.064870 deliveryID 0 before push to receiver - deliveryCount : 0 - linkCredit: 1, len(messages): 0, len(inflight): 0
19:07:51.064908 deliveryID 0 after push to receiver - deliveryCount : 0 - linkCredit: 1, len(messages): 0, len(inflight): 0
19:07:51.064928 deliveryID 0 before exit - deliveryCount : 1 - linkCredit: 0, len(messages): 0
19:07:51.064933 PAUSE Link Mux pause: inflight: 0, credit: 0, deliveryCount: 1, messages: 0, unsettled: 1, maxCredit : 1, settleMode: first
19:07:51.065318 TX: Disposition{Role: Receiver, First: 0, Last: <nil>, Settled: true, State: Accepted, Batchable: false}
```

I don't see any way to resolve this situation starting with v0.13.6. Any idea how this can be resolved?

If I'm reading the previous commits correctly, the empty DeliveryTag is used as a hint for `settleMode == ModeFirst` in `addUnsettled`. Couldn't this instead just use the same `if r.link.receiverSettleMode.value() == ModeSecond` as in `HandleMessage` to be consistent?

Any help would be greatly appreciated.

**Comment by serbrech** [comment]
Mode second means you have to settle the message, it's a "peek lock" mode. 

If you don't settle the message the message is not made available to other subscribers.

It was a bug in the implementation of go-amqp that was not following the spec.

Settlement means calling Accept, or Abandon on the message.
You should always call either accept or abandon in your handlers.

Mode first is when you receive messages "pre-settled"
It is gone from the peer as soon as it is sent, and you effectively don't have to explicitly settle it. That's why it's not being tracked in that case.

Let me know if that clarifies the behavior.

**Comment by serbrech** [comment]
Oh I misunderstood your report.
Hm, so, the broker sends you a message with a 0000 delivery tag in mode first?
I'll have a look at the code and spec again tomorrow.
Thank you for reporting this issue.

**Comment by philschleier** [comment]
That's great, thank you! Yes that is indeed my problem: delivery tag in mode first. Actually, from the debug log of pre-v0.13.6 I can see that the delivery tag increases (I'm guessing that's expected, but only in mode second).

Either way, I've looked at it some more and I still think that if we only call deleteUnsettled if `r.link.receiverSettleMode.value() == ModeSecond` then we should probably use the same condition for addUnsetttled, instead of using the DeliveryTag there.

Once you've had time to look at the spec let me know if that makes sense and I'll happily create a PR for that.

**Comment by serbrech** [comment]
OK, I looked into the code and I can give you a more informed answer now.
I don't see where you see this : 

> my message gets added to the unsettledMessages map because it has a non-empty DeliveryTag

We always add the message to unsettled. there is no condition : https://github.com/Azure/go-amqp/blob/2957005aeaa725796bbd4ae84de0dd5bf58d5fa7/link.go#L512


then in the HandleMessage, ModeSecond requires you to call a disposition. that's the definition of ModeSecond.
and ModeFirst settles it spontaneously.

ModeFirst, spontaneous settling no matter what you do in your handler : https://github.com/Azure/go-amqp/blob/44f772a7887f39e092361f1f4bb4af00b9ea9e65/receiver.go#L54

ModeSecond, we remove the message when you call a settlement func on it in your handling :
https://github.com/Azure/go-amqp/blob/44f772a7887f39e092361f1f4bb4af00b9ea9e65/receiver.go#L33

**Comment by serbrech** [comment]
The logs you show say that your receiver is in settle mode first. the last line shows that a disposition is being sent as expected.
what do you mean by "it breaks our client"?

**Comment by serbrech** [comment]
Ooh, I see this was added by @jhendrixMSFT. I missed it.

https://github.com/Azure/go-amqp/blob/51f7e70071303e8e3e3e867fc068dedeb2f31e44/link.go#L226-L228

This does seem to be making assumption about the sender, and is in fact against the protocol :

> 2.6.12 Transferring A Message
> 
> When an application initiates a message transfer, it assigns a delivery-tag used to track the state of the delivery while the message is in transit. A delivery is considered unsettled at the sender/receiver from the point at which it was sent/received until it has been settled by the sending/receiving application. **_Each delivery MUST be identified by a delivery-tag chosen by the sending application. The delivery-tag MUST be unique amongst all deliveries that could be considered unsettled by either end of the link._** 

I think it was just an optimization and wasn't thought to be a problem.

I think the easiest way to fix this is to remove this condition, and implement more carefully the optimization.

**Comment by jhendrixMSFT** [comment]
Isn't that for mode second though?  I had in fact asked the SB team about this when I made the change.  According to them, messages sent as mode first (pre-settled) don't include a delivery tag.

**Comment by serbrech** [comment]
No, that's the spec
Describing message transfer it does not differ for mode second or mode first.
Seems that Azure Servicebus is not implementing the spec strictly then.
As the bug report shows, other implementation include the tag on mode first.

**Comment by philschleier** [comment]
Nice progress, thanks a lot for looking this stuff up in the spec!

So if I'm understanding you correctly, it is indeed valid to have a delivery tag on the message in mode first and the client should track unsettled messages only in mode second instead of using the delivery tag as a hint for the mode?

You said you don't have proper e2e coverage - is that something you want to change right now? Otherwise I'd open a small PR with that one change if you agree.

**Comment by philschleier** [comment]
Hi @serbrech , I went ahead and created that PR ;-) Let me know what you think.

Regards

**Comment by jhendrixMSFT** [comment]
Fixed in v0.13.9

**Comment by serbrech** [comment]
Just wanted to say thank you @philschleier for taking the time to  file a good issue, debug and even fix it with a PR. Very appreciated, and thanks for your patience too :)
